### ISSUE TITLE
Switch to a more efficient means of building the dependency graph 

### ISSUE THREAD

**Issue by plaird** [salesforce/bazel-eclipse#131]
We currently rely a lot on an aspect to derive the data we need for the dependency graph for each target in each package. It works, but it is really slow.

Ideas:
- Switch to use Bazel query more (would it be faster?)
- Provide an external command line tool to generate the dependency graph into a metafile. User could optionally provide that for import. This would enable just one team member in a repo to take the hit to run it and all coworkers would then have a fast import. (would want to have BEF keep it up to date from there on out)

**Comment by plaird** [comment]
Import was slow but manageable until #104. That PR required us to switch from package level dep info (ran the aspect once per package) to target level dep info (run the aspect per java target per package). For many packages this is now a 5-10x performance degradation as there tends to be a lot of java_test rules.

In #4 we successfully ordered the projects on import such that the classpath need only be computed once per project during import, but that didn't move the import performance needle that much because of this issue.

**Comment by asmittal** [comment]
Though slow, it was working for us. But now its very intermittent. 
Even with latest eclipse, and latest plugin this is what I can observe: 
1) Sometimes the dialogue to select the project to import is shown after 4-5 mintures. 
2) But sometimes even after 2-3 hours no window is shown. Eclipse CPU usage is > 100% .

**Comment by plaird** [comment]
A big issue with using aspects is that they are generated by a **bazel build** on each imported package. If the workspace state is dirty, import can be held up for a long time while the builds get run, especially if big assets like base Docker images need to be downloaded. Because the BEF command line is slightly different, I have found empirically that local disk cache results from non-BEF command line builds aren't used (not sure if this is always the case, but at least in some cases).

**Comment by simontoens** [comment]
@plaird we should figure out how we can make the BEF cmdline args such that they are similar enough to not invalidate any cache.  If I have built on the cmdline and then I run BEF, the BEF build should be very fast because no work should be done by bazel.  Ideally.  What do you think?

**Comment by simontoens** [comment]
Right now I am trying to take bazel out of the picture during import (going with your metafile idea).  I think this will be a useful exercise even if we don't go with that solution in the end.

**Comment by simontoens** [comment]
I am also looking into what IJ is doing.  Probably we should be strongly inspired by IJ.

**Comment by plaird** [comment]
Yes, after we remove the aspects, I think the command line issue goes away 🚀 . We will be doing just plain old 'bazel build xyz' commands. As long as the bazel workspace is correctly configured to be hermetic (host_javabase and javabase configured in .bazelrc) I think then BEF will not need to recompile anything if the workspace state is clean.

**Comment by plaird** [comment]
The other hope I have about this is to analyze the test targets en masse in the package. With bazel query this should be possible. Gather the deps of all java_library targets in the package in one query, gather the deps of all java_test targets in another query...

Then JDT main classpath is the results of the first, the JDT test classpath is the (second - first). Even if the user does not recompute into meta files, this should be reasonably quick.

**Comment by simontoens** [comment]
Just an update:I decided to first refactor the import code because it is a bunch of static methods in a single class and somewhat hard to follow.  The refactoring also helps me understand what is going.  So, progress, just slowly.

**Comment by simontoens** [comment]
@plaird I think what have now is "good enough for the next 6-12 months".  If you agree, let's close this issue.

**Comment by plaird** [comment]
@simontoens yes I agree. Can you write a summary of the work you did in this Issue, and then close.

**Comment by simontoens** [comment]
Summary of changes that improved performance:
- Do not call methods that end up running bazel (loading aspects, querying) in a tight loop, since each bazel invocation, even if bazel doesn't do any work, if expensive. Instead, pass the all arguments into the methods (typically list of targets) and run the methods only once.
- Refactor bazel aspect caching so that when running with a package wildcard target ("\*"), the cache is populated for all concrete targets also.
- Run aspect loading and querying once for all projects being imported, instead of once per project being imported (this doesn't matter if we only import 3 projects, but it has an impact when we import 30+)
